internal/pkg/cli/storage_init.go

```diff
@@ -0,0 +1,235 @@
+// Copyright Amazon.com, Inc or its affiliates. All rights reserved.
+
+package cli
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/config"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/term/color"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/term/log"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/workspace"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+const (
+	dynamoDBStorageType = "Dynamo DB"
+	s3StorageType       = "S3 Bucket"
+)
+
+var storageTypes = []string{
+	dynamoDBStorageType,
+	s3StorageType,
+}
+
+var (
+	fmtStorageInitTypePrompt = "What " + color.Emphasize("type") + " of storage would you like to associate with %s?"
+	storageInitTypeHelp      = "The type of storage you'd like to add to your project. You can choose between key-value (DynamoDB) content (S3) storage types."
+	fmtStorageInitNamePrompt = "What would you like to " + color.Emphasize("name") + " this %s?"
+	storageInitNameHelp      = "The name of this storage resource. You can use the following characters: a-zA-Z0-9-_"
+	storageInitSvcPrompt     = "Which " + color.Emphasize("service") + " would you like to associate with this storage resource?"
+	storageInitSvcHelp       = "The service you'd like to have access to this storage resource. We'll deploy the Cloudformation for the storage when you run `app deploy`."
+)
+
+var fmtStorageInitCreateConfirm = "Okay, we'll create %s %s named %s linked to your %s service."
+
+type s3Config struct {
+	// TODO add website config
+}
+
+type initStorageVars struct {
+	*GlobalOpts
+	StorageType string
+	StorageName string
+	StorageSvc  string
+}
+
+type initStorageOpts struct {
+	initStorageVars
+
+	fs        afero.Fs
+	ws        wsSvcReader
+	store     store
+	appGetter store
+
+	proj *config.Application
+}
+
+func newStorageInitOpts(vars initStorageVars) (*initStorageOpts, error) {
+	store, err := config.NewStore()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't connect to application datastore: %w", err)
+	}
+
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, fmt.Errorf("workspace cannot be created: %w", err)
+	}
+
+	return &initStorageOpts{
+		initStorageVars: vars,
+
+		fs:        &afero.Afero{Fs: afero.NewOsFs()},
+		store:     store,
+		appGetter: store,
+		ws:        ws,
+	}, nil
+}
+
+func (o *initStorageOpts) Validate() error {
+	if o.AppName() == "" {
+		return errNoAppInWorkspace
+	}
+	if o.StorageSvc != "" {
+		if err := o.validateServiceName(); err != nil {
+			return err
+		}
+	}
+
+	if o.StorageType != "" {
+		if err := validateStorageType(o.StorageType); err != nil {
+			return err
+		}
+	}
+
+	if o.StorageName != "" {
+		var err error
+		switch o.StorageType {
+		case dynamoDBStorageType:
+			err = dynamoTableNameValidation(o.StorageName)
+		case s3StorageType:
+			err = s3BucketNameValidation(o.StorageName)
+		default:
+			// use dynamo since it's a superset of s3
+			err = dynamoTableNameValidation(o.StorageName)
```

Comment:
I wonder if this should be an "unexpected storage type" error? so that the next engineer that adds a storage type is forced to implement a name validator explicitly

Reply:
That will error if the user specifies a storage name and not a storage type, though, which is a thorny behavior to parse through. 

If the user specifies 'copilot storage init --name mybucket --svc frontend` are we okay with throwing an error instead of just asking them what type of storage they want to create?

Reply:
ooh i see! Commented about it in the validator file.